bundles/org.openhab.binding.caddx/src/main/java/org/openhab/binding/caddx/internal/CaddxMessage.java

```diff
@@ -0,0 +1,402 @@
+        return message;
+    }
+
+    /**
+     * Returns a string representation of a CaddxMessage.
+     *
+     * @return CaddxMessage string
+     */
+    @Override
+    public String toString() {
+        StringBuilder sb = new StringBuilder();
+
+        CaddxMessageType mt = CaddxMessageType.valueOfMessageType(message[0]);
+        if (mt == null) {
+            return "Unknown message type";
+        }
+
+        sb.append("Message: ");
+        sb.append(String.format("%2s", Integer.toHexString(message[0])));
+        sb.append(" ");
+        sb.append(mt.name);
+        sb.append(System.lineSeparator());
+
+        for (CaddxProperty p : mt.properties) {
+            sb.append("\t").append(p.toString(message));
+            sb.append(System.lineSeparator());
+        }
+
+        return sb.toString();
+    }
+
+    private byte[] getMessageFrameBytesInBinary() {
+        // Calculate bytes
+        // 1 for the startbyte
+        // 1 for the length
+        // 2 for the checksum
+        // n for the count of 0x7d and 0x7e occurrences in the message and checksum
+        int additional = 4;
+        for (int i = 0; i < message.length; i++) {
+            if (message[i] == 0x7d || message[i] == 0x7e) {
+                additional++;
+            }
+        }
+        if (checksum1Calc == 0x7d || checksum1Calc == 0x7e) {
+            additional++;
+        }
+        if (checksum2Calc == 0x7d || checksum2Calc == 0x7e) {
+            additional++;
+        }
+
+        byte[] frame = new byte[message.length + additional];
+        frame[0] = 0x7e;
+        frame[1] = (byte) message.length;
+
+        int fi = 2;
+        for (int i = 0; i < message.length; i++) {
+            byte b = message[i];
+            if (b == 0x7e) {
+                frame[fi++] = 0x7d;
+                b = 0x5e;
+            } else if (b == 0x7d) {
+                frame[fi++] = 0x7d;
+                b = 0x5d;
+            }
+            frame[fi++] = b;
+        }
+
+        // 1st checksum byte
+        if (checksum1Calc == 0x7e) {
+            frame[fi++] = 0x7d;
+            frame[fi++] = 0x5e;
+        } else if (checksum1Calc == 0x7d) {
+            frame[fi++] = 0x7d;
+            frame[fi++] = 0x5d;
+        } else {
+            frame[fi++] = checksum1Calc;
+        }
+        // 2nd checksum byte
+        if (checksum2Calc == 0x7e) {
+            frame[fi++] = 0x7d;
+            frame[fi++] = 0x5e;
+        } else if (checksum2Calc == 0x7d) {
+            frame[fi++] = 0x7d;
+            frame[fi++] = 0x5d;
+        } else {
+            frame[fi++] = checksum2Calc;
+        }
+
+        return frame;
+    }
+
+    private byte[] getMessageFrameBytesInAscii() {
+        // Calculate additional bytes
+        // 1 for the start byte
+        // 2 for the length
+        // 4 for the checksum
+        // 1 for the stop byte
+        int additional = 8;
+
+        int fi = 0;
+        byte[] frame = new byte[2 * message.length + additional];
```

Comment:
What would simplify the code in this method a bunch is if you used a ByteArrayOutputStream here instead.
```
ByteArrayOutputStream frame = new ByteArrayOutputStream(2 * message.length + additional);
frame.write(0x0a);
frame.write(HexUtils.byteToHex(...);
...
return frame.toByteArray();
```

This could also simplify your `getMessageFrameBytesInBinary` method by allowing you to no longer need to calculate the exact frame size beforehand.

Reply:
I'm new to java and don't really know all the specialized classes. So i tend to revert back to basics (plain simple arrays). lol

I will look into it.

Reply:
done